cosmo_tester/test_suites/test_snapshots/snapshots_migration_from_321_and_331_to_34_test.py

```diff
@@ -0,0 +1,276 @@
+            patch.merge_obj(
+                'node_templates.security_group.interfaces',
+                {'cloudify.interfaces.lifecycle': {
+                    'create': {
+                        'inputs': {
+                            'args': {'description': 'hello security group'}
+                        }
+                    }
+                }}
+            )
+
+        activate_path = self._get_activate_path(version)
+        inputs_path = self._prepare_hello_world_inputs()
+
+        arguments = manager_dir, activate_path, hello_blueprint_path, \
+            inputs_path
+
+        self._run_script('install_hello_world.sh', arguments)
+
+    def _uninstall_hello_world(self, version):
+        manager_dir = self._get_manager_dir(version)
+        activate_path = self._get_activate_path(version)
+
+        arguments = manager_dir, activate_path
+
+        self._run_script('uninstall_hello_world.sh', arguments)
+
+    @staticmethod
+    def _get_manager_name(version):
+        return 'manager-{0}'.format(version)
+
+    def _get_manager_dir(self, version):
+        return os.path.join(self.workdir, version.replace('.', ''))
+
+    def _get_venv_dir(self, version):
+        return os.path.join(self._get_manager_dir(version), 'venv')
+
+    def _get_activate_path(self, version):
+        return os.path.join(self._get_venv_dir(version), 'bin', 'activate')
+
+    def _get_snapshots_path(self, version):
+        return os.path.join(self._get_manager_dir(version), 'snapshot')
+
+    def _get_blueprints_dir(self, version):
+        return os.path.join(self._get_manager_dir(version), 'blueprints')
+
+    def _prepare_hello_world_inputs(self):
+        inputs = {
+            'agent_user': self.env.cloudify_agent_user,
+            'image': self.env.ubuntu_trusty_image_name,
+            'flavor': self.env.flavor_name
+        }
+
+        with tempfile.NamedTemporaryFile(delete=False) as f:
+            f.write(json.dumps(inputs))
+
+            return f.name
+
+    def _run_script(self, name, arguments):
+        script_path = pkg_resources.resource_filename(
+            cosmo_tester.__name__,
+            'resources/scripts/snapshots_migration_scripts/{0}'.format(
+                name)
+        )
+
+        command = 'bash "{0}"'.format(script_path)
+        arguments = ' '.join('"{}"'.format(a) for a in arguments)
+
+        rc = subprocess.call(' '.join([command, arguments]), shell=True)
+
+        if rc:
+            self.fail('Running script {0} failed (return code = {1})'.format(
+                name, rc))
+
+    def _test_migration(self, source, destination):
+        self._bootstrap_manager(source)
+        self._bootstrap_manager(destination)
+        self._install_hello_world(source)
+        self._create_and_download_snapshot(source)
+        self._upload_snapshot(source, destination)
+        self._restore_snapshot_and_install_agents(destination)
+        self._uninstall_hello_world(destination)
```

Comment:
calling uninstall alone is not sufficient to verify that everything works because the uninstall workflow will silently ignore errors. I suggest maybe using the "execute_operation" workflow to "stop" the webserver and assert it is stopped. (it should be verified beforehand that it does respond). What is left untested is the openstack plugin (or more generally, the manager plugins that are installed as part of this process. I'm not sure what exactly to do here but something should probably be done